cpp/src/parquet/column_reader.cc

```diff
@@ -1370,6 +1402,54 @@ class TypedRecordReader : public TypedColumnReaderImpl<DType>,
     return bytes_for_values;
   }
 
+  // Two parts different from original HasNextInternal:
+  //   1. When reading a new page, the corresponding skip info will be updated;
+  //   2. When current page has more data to read, it will check skip info
+  //   inside current page;
+  bool HasNextInternal() override {
+    // Either there is no data page available yet, or the data page has been
+    // exhausted
+    if (this->num_buffered_values_ == 0 ||
+        this->num_decoded_values_ == this->num_buffered_values_) {
+      if (!this->ReadNewPage() || this->num_buffered_values_ == 0) {
+        return false;
+      }
+
+      // Skip header rows and hide last rows for this page
+      auto read_states = this->pager_->GetPageReadStates();
+      if (read_states != nullptr) {
+        PageSkipInfo& skip_info = read_states->NextPageSkipInfo();
+        const int64_t records_to_skip = skip_info.SkipRowNum();
+        const int64_t skipped_records = SkipRecords(records_to_skip);
```

Comment:
IMO, doing `SkipRecords` in `HasNextInternal` would be a bit tricky...

Reply:
> IMO, doing `SkipRecords` in `HasNextInternal` would be a bit tricky...

Maybe we can implement the row skipping logic in ``ReadRecords``.

Reply:
do we need to document clearly how (or if?) the [data page filter](https://github.com/apache/arrow/blob/main/cpp/src/parquet/column_reader.cc#L474) and the page pruning can work with each other??

Reply:
Is data_page_filter used in this patch?

Reply:
the `ReadNextPage` may use data page filter. Assuming users set both filter and row ranges, and some row ranges overlaps with pages that can be skipped by the filter. Would `SkipRecords(records_to_skip)` skip wrong records?

Reply:
Seems multiple machanism is applied here...

Reply:
@jp0317 Should we combine `ShouldSkipPage` with PageIndex? Our current codebase relies something weird here. 🤔

Reply:
> do we need to document clearly how (or if?) the [data page filter](https://github.com/apache/arrow/blob/main/cpp/src/parquet/column_reader.cc#L474) and the page pruning can work with each other??

``data_page_filter`` is used to filter out non-hit pages. In the current implementation, ``SkipRecords`` is used to skip unnecessary rows within the hit pages, they do not conflict with each other. Using with ``data_page_filter`` is not currently supported, but such scenarios can be supported with very minor modifications to ``NextPage``.

Reply:
> @jp0317 Should we combine `ShouldSkipPage` with PageIndex? Our current codebase relies something weird here. 🤔

It seems`ShouldSkipPage` [targets page header statistics](https://github.com/apache/arrow/blob/main/cpp/src/parquet/column_reader.cc#L299-L301)..iiuc we might want two version of `skipping page` (rather than combining into one): one based on PageIndex before actually IO on that page, one based on page header after IO is done. 

I think we'd need more clearer docs on those skip-related apis as currently it seems they are supposed to be used in a certain way but not stated in the doc..